### PR TITLE
customise fire button to preserve logins

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/72649045549333/1139199704136673
Tech Design URL:

* https://app.asana.com/0/72649045549333/1153240009748114
* https://app.asana.com/0/72649045549333/1154411222422091

Product Review: https://app.asana.com/0/59792373528535/1159798310046168

CC:

**Description**:

Allows users on iOS 13 to keep cookies for sites they want to stay logged in to.

Note that login detection does not work for all sites, e.g. Twitter, but does work for many, e.g. Reddit, Facebook.   Our intention is to get good, but not necessarily perfect coverage.  While testing, if you find a site that this doesn't work on which seems important please comment on this task: 

https://app.asana.com/0/72649045549333/1159798310046161

**Code review tips**:

Check out the code and run it, going through at least the tests listed below plus more.  Check for:

* bugs
* crashes
* memory leaks
* inconsistent behaviour across devices (at least iOS 12 and 13 on a phone and tablet)

In the code look for:
* inconsistencies compared to the original code base
* wrong file headers
* dubious logic 
* obvious optimisations 
* anything that needs further explanation (e.g. it's not clear why the code is doing what it is doing)

**Steps to test this PR**:

***Upgrades***

1. Upgrade from an existing installation.
1. Expected: App should launch and function as expected
- [x] Pass

--
1. Navigate to settings
1. Expected: Keep Me Signed In should be **Off**.
- [x] Pass

--
1. Browse to a website (e.g. reddit), login and then hit the fire button.
1. Expected: No prompt and navigating back to Reddit should require you to login in again
- [x] Pass

***New installs***

1. Fresh install the app
1. Browse to any website but do not login, then execute fire button
1. Expected: No prompt
- [x] Pass

--
1. Browse to a website (eg Reddit), login then execute fire button.
1. Expected: Should see "Keep Me Signed In?" prompt
1. Select "Keep Me Signed In"
1. Browse back to the website you logged in to
1. Expected: You should still be logged in
1. Navigate to Settings
1. Expected: Keep Me Signed In should be **On**.
- [x] Pass

--
1. Browse to a website (eg Reddit), login then execute fire button.
1. Expected: Should NOT see "Keep Me Signed In?" prompt
- [x] Pass

-- 
1. Fresh install
1. Browse to a website (eg Reddit), login then execute fire button.
1. Expected: Should see "Keep Me Signed In?" prompt
1. Select "Sign Me Out"
1. Browse back to the website you logged in to
1. Expected: You should NOT be logged in
1. Navigate to Settings
1. Expected: Keep Me Signed In should be **Off**.
- [x] Pass

***Settings***

1. Settings > Keep Me Signed In toggle should reflect the current state of the functionality:
* Enabled/Disabled
* List of currently logged in domains
1. Delete individual domains with a swipe
1. Use edit to Clear all domains
1. Turn switch off to be prompted to clear all domains, but only if there are domains to clear
- [x] Pass

***< iOS 13***

1. Testing on a device < iOS 13
1. Expected: The user does not see this feature
- [x] Pass

***Pixels***

Ensure pixels are sent as per https://app.asana.com/0/72649045549333/1155929926923302
- [x] Pass

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/PreserveLogins.swift [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                allowedDomains += detectedDomains
+                detectedDomains = []
+            } else {
+                detectedDomains = allowedDomains
+                allowedDomains = []
+            }
+        }
+    }
+    
+    public var prompted: Bool {
```
Comment (tomasstrba) [inline Core/PreserveLogins.swift:80]: In cases like this UserDefault property wrapper could be helpful for us. We wouldn't repeat same code again and make it more compact. Maybe it is good candidate for LHF days.

File: Core/Pixel.swift [hunk 2]
```diff
@@ -180,6 +180,18 @@ public enum PixelName: String {
     
     case brokenSiteReported = "m_bsr"
 
+    case preserveLoginsUserDecisionPreserve = "m_pl_p"
+    case preserveLoginsUserDecisionForget = "m_pl_f"
+    case preserveLoginsSettingsWhilePreserving = "m_pl_s_p"
+    case preserveLoginsSettingsWhileForgetting = "m_pl_s_f"
```
Comment (tomasstrba) [inline Core/Pixel.swift:186]: Should "preserveLoginsSettingsWhileForgetting" pixel be fired somewhere in code or did you forget to remove definition?

**Review by tomasstrba — CHANGES_REQUESTED** [review]
@brindy good job! This is really handy feature

There are just two things I found:
- preserveLoginsSettingsWhileForgetting is not used anywhere
- I found exception on iPad Air 2, iOS 12. Go to Settings > Keep Me Signed, Use edit to Clear all domains, Tap on ''Clear data & Sign Out

**Comment by brindy** [comment]
Thanks @tomasstrba! 

I am getting some feedback from the product review as well so will ping you when I need another review. 

The property wrapper is a great idea - I'll see about adding that, then we can clean up a lot of code throughout the codebase at some point in the future.

Thanks for the heads up about the crash - will investigate, though I suspect product review changes will actually make that go away.  Will ping you soon.

File: Core/PreserveLogins.swift (hunk 1 above)
Comment (brindy) [reply Core/PreserveLogins.swift:80]: I had a quick look, but it makes testing quite difficult as we tend to inject a user defaults object in that has a known state.  Nice idea, but I think I need to ponder it more before putting it in.

**Comment by brindy** [comment]
Hey @tomasstrba - I *think* that's it for copy.  We also decided to restrict this functionality to iOS 13 users, so have added another test for that.  Ready to review again when you get chance.  Thanks!

**Comment by tomasstrba** [comment]
LGTM! Good job @brindy!

**Comment by brindy** [comment]
Thanks @tomasstrba !

**Pull request merged**